// @umituz/react-native-filesystem/src/infrastructure/services/download.service.ts
/**
 * Download Service
 * Single Responsibility: Download files from URLs with progress tracking
 */

import { File, Paths, Directory } from "expo-file-system";
import type { FileOperationResult, DownloadProgress } from "../../domain/entities/File";
import { FileUtils } from "../../domain/entities/File";

/** Progress callback type */
export type DownloadProgressCallback = (progress: DownloadProgress) => void;

/** Download with progress result */
export interface DownloadWithProgressResult extends FileOperationResult {
  fromCache?: boolean;
}

/**
 * Download file from URL (simple version without progress)
 */
export async function downloadFile(
  url: string,
  destinationUri?: string,
): Promise<FileOperationResult> {
  try {
    let destination: File | typeof Paths.document;

    if (destinationUri) {
      destination = new File(destinationUri);
    } else {
      const filename = FileUtils.generateUniqueFilename("download");
      destination = new File(Paths.document, filename);
    }

    const result = await File.downloadFileAsync(url, destination, {
      idempotent: true,
    });
    return { success: true, uri: result.uri };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : "Download failed",
    };
  }
}

/**
 * Generate hash from URL for cache filename
 */
function hashUrl(url: string): string {
  let hash = 0;
  for (let i = 0; i < url.length; i++) {
    const char = url.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash = hash & hash;
  }
  return Math.abs(hash).toString(36);
}

/**
 * Get extension from URL
 */
function getExtensionFromUrl(url: string): string {
  const urlPath = url.split("?")[0];
  const ext = urlPath.split(".").pop()?.toLowerCase() || "mp4";
  const validExts = ["mp4", "mov", "m4v", "webm", "jpg", "png", "pdf"];
  return validExts.includes(ext) ? ext : "mp4";
}

/**
 * Download file with progress tracking (uses expo/fetch for progress)
 * Falls back to simple download if progress not needed
 */
export async function downloadFileWithProgress(
  url: string,
  cacheDir: string,
  onProgress?: DownloadProgressCallback,
): Promise<DownloadWithProgressResult> {
  try {
    // Ensure cache directory exists
    const dir = new Directory(cacheDir);
    if (!dir.exists) {
      dir.create({ intermediates: true, idempotent: true });
    }

    // Generate filename from URL hash
    const hash = hashUrl(url);
    const ext = getExtensionFromUrl(url);
    const filename = `cached_${hash}.${ext}`;
    const destUri = FileUtils.joinPaths(cacheDir, filename);

    // Check if already cached
    const cachedFile = new File(destUri);
    if (cachedFile.exists && (cachedFile.size ?? 0) > 0) {
      return { success: true, uri: destUri, fromCache: true };
    }

    // Download using fetch with progress (expo/fetch)
    const response = await fetch(url);

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    const contentLength = response.headers.get("content-length");
    const totalBytes = contentLength ? parseInt(contentLength, 10) : 0;

    if (!response.body) {
      // Fallback: no streaming support, use simple download
      const result = await downloadFile(url, destUri);
      return { ...result, fromCache: false };
    }

    // Stream download with progress
    const reader = response.body.getReader();
    const chunks: Uint8Array[] = [];
    let receivedBytes = 0;

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      chunks.push(value);
      receivedBytes += value.length;

      if (onProgress) {
        onProgress({
          totalBytesWritten: receivedBytes,
          totalBytesExpectedToWrite: totalBytes || receivedBytes,
        });
      }
    }

    // Combine chunks and write to file
    const allChunks = new Uint8Array(receivedBytes);
    let position = 0;
    for (const chunk of chunks) {
      allChunks.set(chunk, position);
      position += chunk.length;
    }

    // Write bytes to file
    const file = new File(destUri);
    file.write(allChunks);

    return { success: true, uri: destUri, fromCache: false };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : "Download failed",
    };
  }
}

/**
 * Check if URL is cached
 */
export function isUrlCached(url: string, cacheDir: string): boolean {
  try {
    const hash = hashUrl(url);
    const ext = getExtensionFromUrl(url);
    const filename = `cached_${hash}.${ext}`;
    const destUri = FileUtils.joinPaths(cacheDir, filename);
    const file = new File(destUri);
    return file.exists && (file.size ?? 0) > 0;
  } catch {
    return false;
  }
}

/**
 * Get cached file URI if exists
 */
export function getCachedFileUri(url: string, cacheDir: string): string | null {
  try {
    const hash = hashUrl(url);
    const ext = getExtensionFromUrl(url);
    const filename = `cached_${hash}.${ext}`;
    const destUri = FileUtils.joinPaths(cacheDir, filename);
    const file = new File(destUri);
    if (file.exists && (file.size ?? 0) > 0) {
      return destUri;
    }
    return null;
  } catch {
    return null;
  }
}

/**
 * Delete cached file
 */
export function deleteCachedFile(url: string, cacheDir: string): boolean {
  try {
    const hash = hashUrl(url);
    const ext = getExtensionFromUrl(url);
    const filename = `cached_${hash}.${ext}`;
    const destUri = FileUtils.joinPaths(cacheDir, filename);
    const file = new File(destUri);
    if (file.exists) {
      file.delete();
    }
    return true;
  } catch {
    return false;
  }
}
