// @umituz/react-native-filesystem/src/infrastructure/services/FileSystemService.ts
/**
 * FileSystem Service - Facade
 * Delegates to specialized services following SOLID principles
 */

import { readFile, readFileAsBase64 } from "./file-reader.service";
import { writeFile } from "./file-writer.service";
import { deleteFile, copyFile, moveFile } from "./file-manager.service";
import {
  createDirectory,
  listDirectory,
  getDirectoryPath,
  getDocumentDirectory,
  getCacheDirectory,
} from "./directory.service";
import { getFileInfo, fileExists, getFileSize } from "./file-info.service";
import { downloadFile } from "./download.service";
import { clearCache, getDirectorySize } from "./cache.service";
import { generateFilePath } from "./file-path.service";
import { FileUtils } from "../../domain/entities/File";
import type {
  FileEncoding,
  DirectoryType,
  FileOperationResult,
} from "../../domain/entities/File";

/**
 * FileSystem Service - Clean facade for all file operations
 * Delegates to specialized services following Single Responsibility Principle
 */
export class FileSystemService {
  // File Reading
  static async readFile(
    uri: string,
    encoding: FileEncoding = "utf8"
  ): Promise<string | null> {
    return readFile(uri, encoding);
  }

  static async readFileAsBase64(uri: string): Promise<string | null> {
    return readFileAsBase64(uri);
  }

  // File Writing
  static async writeFile(
    uri: string,
    content: string,
    encoding: FileEncoding = "utf8"
  ): Promise<FileOperationResult> {
    return writeFile(uri, content, encoding);
  }

  // File Management
  static async deleteFile(uri: string): Promise<boolean> {
    return deleteFile(uri);
  }

  static async copyFile(
    sourceUri: string,
    destinationUri: string
  ): Promise<FileOperationResult> {
    return copyFile(sourceUri, destinationUri);
  }

  static async moveFile(
    sourceUri: string,
    destinationUri: string
  ): Promise<FileOperationResult> {
    return moveFile(sourceUri, destinationUri);
  }

  // Directory Operations
  static async createDirectory(uri: string): Promise<boolean> {
    return createDirectory(uri);
  }

  static async listDirectory(uri: string): Promise<string[]> {
    return listDirectory(uri);
  }

  static getDirectoryPath(type: DirectoryType): string {
    return getDirectoryPath(type);
  }

  static getDocumentDirectory(): string {
    return getDocumentDirectory();
  }

  static getCacheDirectory(): string {
    return getCacheDirectory();
  }

  // File Information
  static async getFileInfo(uri: string) {
    return getFileInfo(uri);
  }

  static async exists(uri: string): Promise<boolean> {
    return fileExists(uri);
  }

  static async getFileSize(uri: string): Promise<number> {
    return getFileSize(uri);
  }

  // Downloads
  static async downloadFile(
    url: string,
    destinationUri?: string
  ): Promise<FileOperationResult> {
    return downloadFile(url, destinationUri);
  }

  // Cache Management
  static async clearCache(): Promise<boolean> {
    return clearCache();
  }

  static async getDirectorySize(uri: string): Promise<number> {
    return getDirectorySize(uri);
  }

  // File Path Generation
  static generateFilePath(
    filename: string,
    directory: DirectoryType = "documentDirectory"
  ): string {
    return generateFilePath(filename, directory);
  }

  // Convenience methods
  static async copyToCache(
    sourceUri: string,
    filename?: string
  ): Promise<FileOperationResult> {
    const cacheDir = getCacheDirectory();
    const uniqueFilename = FileUtils.generateUniqueFilename(
      filename || sourceUri.split("/").pop() || "file"
    );
    const destinationUri = FileUtils.joinPaths(cacheDir, uniqueFilename);
    return copyFile(sourceUri, destinationUri);
  }

  static async copyToDocuments(
    sourceUri: string,
    filename?: string
  ): Promise<FileOperationResult> {
    const docDir = getDocumentDirectory();
    const uniqueFilename = FileUtils.generateUniqueFilename(
      filename || sourceUri.split("/").pop() || "file"
    );
    const destinationUri = FileUtils.joinPaths(docDir, uniqueFilename);
    return copyFile(sourceUri, destinationUri);
  }
}
