// @umituz/react-native-design-system/src/responsive/index.ts
/**
 * @umituz/react-native-design-system/responsive - Public API
 *
 * Responsive sizing and layout utilities for React Native.
 * For device detection, use '@umituz/react-native-design-system/device'.
 */

// Hook exports
export { useResponsive } from './useResponsive';
export type { UseResponsiveReturn } from './useResponsive';

// Responsive sizing utilities
export {
  getResponsiveLogoSize,
  getResponsiveInputHeight,
  getResponsiveHorizontalPadding,
  getResponsiveVerticalPadding,
  getScreenLayoutConfig,
  getResponsiveBottomPosition,
  getResponsiveFABPosition,
  getResponsiveTabBarHeight,
  getResponsiveTabBarConfig,
  type ResponsiveTabBarConfig,
  type ScreenLayoutConfig,
  getResponsiveModalMaxHeight,
  getResponsiveMinModalHeight,
  getResponsiveModalWidth,
  getResponsiveModalHeight,
  getResponsiveModalBorderRadius,
  getResponsiveModalMaxWidth,
  getResponsiveBackdropOpacity,
  getResponsiveModalLayout,
  getResponsiveBottomSheetLayout,
  getResponsiveDialogLayout,
  type ResponsiveModalLayout,
  type ResponsiveBottomSheetLayout,
  type ResponsiveDialogLayout,
  getResponsiveIconContainerSize,
  getResponsiveGridColumns,
  getResponsiveGridCellSize,
  type GridCellSizeConfig,
  getResponsiveMaxWidth,
  getResponsiveFontSize,
  MODAL_CONFIG,
} from './responsive';

// Platform constants
export {
  IOS_HIG,
  PLATFORM_CONSTANTS,
  isValidTouchTarget,
  getMinTouchTarget,
} from './platformConstants';

// Config exports
export { DEVICE_BREAKPOINTS } from './config';
