// @umituz/react-native-haptics/src/infrastructure/services/HapticService.ts
/**
 * Haptics Domain - Haptic Service
 *
 * Service for haptic feedback using expo-haptics.
 * Provides abstraction layer for vibration and feedback.
 *
 * @domain haptics
 * @layer infrastructure/services
 */

import * as Haptics from 'expo-haptics';
import type { ImpactStyle, NotificationType, HapticPattern } from '../../domain/entities/Haptic';


/**
 * Haptic feedback service
 */
export class HapticService {
  /**
   * Trigger impact feedback (Light, Medium, Heavy)
   */
  static async impact(style: ImpactStyle = 'Light'): Promise<void> {
    try {
      await Haptics.impactAsync(
        style === 'Light' ? Haptics.ImpactFeedbackStyle.Light :
        style === 'Medium' ? Haptics.ImpactFeedbackStyle.Medium :
        Haptics.ImpactFeedbackStyle.Heavy
      );
    } catch (error) {
      // Silent fail - haptics not critical
    }
  }

  /**
   * Trigger notification feedback (Success, Warning, Error)
   */
  static async notification(type: NotificationType): Promise<void> {
    try {
      await Haptics.notificationAsync(
        type === 'Success' ? Haptics.NotificationFeedbackType.Success :
        type === 'Warning' ? Haptics.NotificationFeedbackType.Warning :
        Haptics.NotificationFeedbackType.Error
      );
    } catch (error) {
      // Silent fail
    }
  }

  /**
   * Trigger selection feedback (for pickers, sliders)
   */
  static async selection(): Promise<void> {
    try {
      await Haptics.selectionAsync();
    } catch (error) {
      // Silent fail
    }
  }

  /**
   * Trigger haptic pattern
   */
  static async pattern(pattern: HapticPattern): Promise<void> {
    try {
      if (pattern === 'selection') {
        await HapticService.selection();
        return;
      }

      // Success, Warning, Error map to notification feedback
      if (pattern === 'Success' || pattern === 'Warning' || pattern === 'Error') {
        await HapticService.notification(pattern as NotificationType);
        return;
      }

      // Default to light impact
      await HapticService.impact('Light');
    } catch (error) {
      // Silent fail
    }
  }

  /**
   * Common haptic patterns (convenience methods)
   */
  static async buttonPress(): Promise<void> {
    await HapticService.impact('Light');
  }

  static async success(): Promise<void> {
    await HapticService.pattern('Success');
  }

  static async error(): Promise<void> {
    await HapticService.pattern('Error');
  }

  static async warning(): Promise<void> {
    await HapticService.pattern('Warning');
  }

  static async delete(): Promise<void> {
    await HapticService.impact('Medium');
  }

  static async refresh(): Promise<void> {
    await HapticService.impact('Light');
  }

  static async selectionChange(): Promise<void> {
    await HapticService.pattern('selection');
  }

  static async longPress(): Promise<void> {
    await HapticService.impact('Medium');
  }
}
