// @umituz/react-native-design-system/src/theme/core/colors/LightColors.ts
/**
 * LIGHT THEME COLORS
 * 
 * Light theme color palette with warm orange harmony
 */

export const lightColors = {
  // PRIMARY BRAND COLORS - Warm Orange & Harmony Bloom
  primary: '#FF6B35',          // Vibrant Orange
  primaryLight: '#FF8C42',     // Warm Orange
  primaryDark: '#FF4500',      // Orange Red

  secondary: '#FFB88C',        // Soft Peach
  secondaryLight: '#FFCC99',   // Light Peach
  secondaryDark: '#FF8C69',    // Salmon

  accent: '#FFB6C1',           // Light Pink (Bloom)
  accentLight: '#FFA07A',      // Light Salmon
  accentDark: '#FF8C69',       // Salmon

  // MATERIAL DESIGN 3 - ON COLORS (Text on colored backgrounds)
  onPrimary: '#FFFFFF',           // Text on primary background
  onSecondary: '#FFFFFF',         // Text on secondary background
  onSuccess: '#FFFFFF',           // Text on success background
  onError: '#FFFFFF',             // Text on error background
  onWarning: '#000000',           // Text on warning background
  onInfo: '#FFFFFF',              // Text on info background
  onSurface: '#1E293B',           // Text on surface
  onBackground: '#1E293B',        // Text on background
  onSurfaceDisabled: '#CBD5E1',   // Disabled text color
  onSurfaceVariant: '#64748B',    // Text on surface variant

  // MATERIAL DESIGN 3 - CONTAINER COLORS (Lighter versions for containers)
  primaryContainer: '#FFE4CD',    // Light orange container
  onPrimaryContainer: '#CC4A1F',  // Text on primary container
  secondaryContainer: '#FFF8DC',  // Light peach container
  onSecondaryContainer: '#CC8C5F', // Text on secondary container
  errorContainer: '#FEE2E2',      // Light container using error
  onErrorContainer: '#991B1B',    // Text on error container

  // MATERIAL DESIGN 3 - OUTLINE
  outline: '#CBD5E1',             // Default outline color
  outlineVariant: '#E2E8F0',      // Lighter outline variant
  outlineDisabled: '#E2E8F0',     // Disabled outline color

  // SEMANTIC UI COLORS
  success: '#10B981',
  successLight: '#34D399',
  successDark: '#059669',

  error: '#EF4444',
  errorLight: '#F87171',
  errorDark: '#DC2626',

  warning: '#F59E0B',
  warningLight: '#FBBF24',
  warningDark: '#D97706',

  info: '#FF8C42',              // Warm Orange for info
  infoLight: '#FFA07A',         // Light Salmon
  infoDark: '#FF6347',          // Tomato

  // SEMANTIC CONTAINER COLORS (Light mode)
  successContainer: '#D1FAE5',        // Light container for success states
  onSuccessContainer: '#065F46',      // Text on success container
  warningContainer: '#FEF3C7',        // Light container for warning states
  onWarningContainer: '#92400E',      // Text on warning container
  infoContainer: '#FFE4CD',           // Light orange container for info states
  onInfoContainer: '#CC4A1F',         // Text on info container

  // GRAYSCALE PALETTE
  gray50: '#FAFAFA',
  gray100: '#F4F4F5',
  gray200: '#E4E4E7',
  gray300: '#D4D4D8',
  gray400: '#A1A1AA',
  gray500: '#71717A',
  gray600: '#52525B',
  gray700: '#3F3F46',
  gray800: '#27272A',
  gray900: '#18181B',

  // BACKGROUND COLORS
  backgroundPrimary: '#FFFFFF',
  backgroundSecondary: '#F8FAFC',

  surface: '#FFFFFF',
  surfaceVariant: '#F1F5F9',
  surfaceSecondary: '#F1F5F9', // Alias
  surfaceDisabled: '#F4F4F5',     // Disabled surface color

  // TEXT COLORS
  textPrimary: '#1E293B',
  textSecondary: '#64748B',
  textTertiary: '#94A3B8',
  textDisabled: '#CBD5E1',
  textInverse: '#FFFFFF',

  // BORDER COLORS
  border: '#E2E8F0',
  borderLight: '#F1F5F9',
  borderMedium: '#CBD5E1',
  borderFocus: '#3B82F6',
  borderDisabled: '#F1F5F9',

  // COMPONENT-SPECIFIC COLORS
  buttonPrimary: '#FF6B35',        // Vibrant Orange
  buttonSecondary: '#FFB88C',      // Soft Peach

  inputBackground: '#FFFFFF',
  inputBorder: '#E2E8F0',

  cardBackground: '#FFFFFF',

  // COLOR ALIASES (for backward compatibility and convenience)
  text: '#1E293B',                    // Alias for textPrimary
  background: '#FFFFFF',              // Alias for backgroundPrimary
  card: '#FFFFFF',                    // Alias for cardBackground

  // SPECIAL COLORS
  transparent: 'transparent',
  black: '#000000',
  white: '#FFFFFF',

  // RGBA OVERLAY COLORS (for modals, cards, etc.)
  modalOverlay: 'rgba(0, 0, 0, 0.5)',
  overlaySubtle: 'rgba(0, 0, 0, 0.05)',
  overlayLight: 'rgba(0, 0, 0, 0.1)',
  overlayMedium: 'rgba(0, 0, 0, 0.3)',
  overlayBackground: 'rgba(0, 0, 0, 0.05)',

  whiteOverlay: 'rgba(255, 255, 255, 0.2)',
  whiteOverlayStrong: 'rgba(255, 255, 255, 0.95)',
  whiteOverlayBorder: 'rgba(255, 255, 255, 0.5)',

  textWhiteOpacity: 'rgba(255, 255, 255, 0.8)',

  errorBackground: 'rgba(239, 68, 68, 0.1)',
  primaryBackground: 'rgba(255, 107, 53, 0.1)',  // Orange background

  cardOverlay: 'rgba(0, 0, 0, 0.15)',

  inputBackground_RGBA: 'rgba(248, 250, 252, 0.9)',

  // GRADIENTS - Warm Orange Harmony
  gradient: ['#FF6B35', '#FFB88C'],  // Orange to Peach gradient
};
